File: modules/collab/test/integration/rbac.test.mjs
// RBAC resolver: roles from token claims + tenant isolation by room prefix. Pure over a `user` object,
// so no tokens/sockets needed. Plain ESM, run by vitest.

import { describe, expect, it } from "vitest";
import { canWrite, createClaimsRoleResolver } from "../../server/rbac.mjs";

const authorizeRoom = createClaimsRoleResolver();

describe("RBAC — role resolution", () => {
  it("grants full access to an unauthenticated user (auth off / local dev)", () => {
    expect(authorizeRoom({ user: null, room: "playground" })).toBe("editor");
  });

  it("defaults an authenticated user with no roles claim to editor", () => {
    const user = { sub: "u", tenant: null, roles: null };
    expect(authorizeRoom({ user, room: "anything" })).toBe("editor");
  });

  it("returns the per-room role from the claim (by full id or bare id)", () => {
    const user = { sub: "u", tenant: null, roles: { "acme/board1": "viewer", board2: "owner" } };
    expect(authorizeRoom({ user, room: "acme/board1" })).toBe("viewer");
    expect(authorizeRoom({ user, room: "board2" })).toBe("owner");
  });

  it("denies a room the roles claim doesn't list (when a claim is present)", () => {
    const user = { sub: "u", tenant: null, roles: { board1: "editor" } };
    expect(authorizeRoom({ user, room: "board9" })).toBeNull();
  });

  it("rejects an unknown role value", () => {
    const user = { sub: "u", tenant: null, roles: { board1: "superadmin" } };
    expect(authorizeRoom({ user, room: "board1" })).toBeNull();
  });
});

describe("RBAC — tenant isolation", () => {
  it("admits a room namespaced to the user's tenant", () => {
    const user = { sub: "u", tenant: "org_acme", roles: { "org_acme/board1": "editor" } };
    expect(authorizeRoom({ user, room: "org_acme/board1" })).toBe("editor");
  });

  it("denies a room in another tenant, even with a matching role key", () => {
    const user = { sub: "u", tenant: "org_acme", roles: { "org_evil/board1": "owner" } };
    expect(authorizeRoom({ user, room: "org_evil/board1" })).toBeNull();
  });
});

describe("RBAC — canWrite", () => {
  it("owner and editor can write; viewer cannot", () => {
    expect(canWrite("owner")).toBe(true);
    expect(canWrite("editor")).toBe(true);
    expect(canWrite("viewer")).toBe(false);
    expect(canWrite(null)).toBe(false);
  });
});

File: modules/relay/relay/rbac_test.go
// Ported from modules/collab/test/integration/rbac.test.mjs.
package relay

import "testing"

func TestRBACGrantsFullAccessToUnauthenticatedUser(t *testing.T) {
	authorizeRoom := NewClaimsRoleResolver("")
	if got := authorizeRoom(nil, "playground"); got != RoleEditor {
		t.Errorf("got %q, want editor", got)
	}
}

func TestRBACFailsClosedByDefault(t *testing.T) {
	authorizeRoom := NewClaimsRoleResolver("")
	user := &User{Sub: "u"}
	if got := authorizeRoom(user, "anything"); got != "" {
		t.Errorf("got %q, want no access", got)
	}
}

func TestRBACDevPostureDefaultEditor(t *testing.T) {
	devResolver := NewClaimsRoleResolver(RoleEditor)
	user := &User{Sub: "u"}
	if got := devResolver(user, "anything"); got != RoleEditor {
		t.Errorf("got %q, want editor", got)
	}
	explicit := &User{Sub: "u", Roles: map[string]Role{"x": RoleOwner}}
	if got := devResolver(explicit, "x"); got != RoleOwner {
		t.Errorf("got %q, want owner (explicit grant under dev default)", got)
	}
}

func TestRBACCanFailClosedButHonoursExplicitGrants(t *testing.T) {
	denyByDefault := NewClaimsRoleResolver("")
	user := &User{Sub: "u"}
	if got := denyByDefault(user, "anything"); got != "" {
		t.Errorf("got %q, want no access", got)
	}
	explicit := &User{Sub: "u", Roles: map[string]Role{"x": RoleEditor}}
	if got := denyByDefault(explicit, "x"); got != RoleEditor {
		t.Errorf("got %q, want editor", got)
	}
}

func TestRBACPerRoomRoleByFullOrBareID(t *testing.T) {
	authorizeRoom := NewClaimsRoleResolver("")
	user := &User{Sub: "u", Roles: map[string]Role{"acme/board1": RoleViewer, "board2": RoleOwner}}
	if got := authorizeRoom(user, "acme/board1"); got != RoleViewer {
		t.Errorf("got %q, want viewer", got)
	}
	if got := authorizeRoom(user, "board2"); got != RoleOwner {
		t.Errorf("got %q, want owner", got)
	}
}

func TestRBACDeniesUnlistedRoom(t *testing.T) {
	authorizeRoom := NewClaimsRoleResolver("")
	user := &User{Sub: "u", Roles: map[string]Role{"board1": RoleEditor}}
	if got := authorizeRoom(user, "board9"); got != "" {
		t.Errorf("got %q, want no access", got)
	}
}

func TestRBACRejectsUnknownRoleValue(t *testing.T) {
	authorizeRoom := NewClaimsRoleResolver("")
	user := &User{Sub: "u", Roles: map[string]Role{"board1": Role("superadmin")}}
	if got := authorizeRoom(user, "board1"); got != "" {
		t.Errorf("got %q, want no access", got)
	}
}

func TestRBACTenantIsolationAdmitsMatchingTenant(t *testing.T) {
	authorizeRoom := NewClaimsRoleResolver("")
	user := &User{Sub: "u", Tenant: "org_acme", Roles: map[string]Role{"org_acme/board1": RoleEditor}}
	if got := authorizeRoom(user, "org_acme/board1"); got != RoleEditor {
		t.Errorf("got %q, want editor", got)
	}
}

func TestRBACTenantIsolationDeniesOtherTenant(t *testing.T) {
	authorizeRoom := NewClaimsRoleResolver("")
	user := &User{Sub: "u", Tenant: "org_acme", Roles: map[string]Role{"org_evil/board1": RoleOwner}}
	if got := authorizeRoom(user, "org_evil/board1"); got != "" {
		t.Errorf("got %q, want no access despite a matching role key", got)
	}
}

func TestCanWrite(t *testing.T) {
	if !CanWrite(RoleOwner) {
		t.Error("owner should write")
	}
	if !CanWrite(RoleEditor) {
		t.Error("editor should write")
	}
	if CanWrite(RoleViewer) {
		t.Error("viewer should not write")
	}
	if CanWrite("") {
		t.Error("no role should not write")
	}
}
